router/batchrouter/asyncdestinationmanager/yandexmetrica/yandexmetrica.go

```diff
@@ -0,0 +1,382 @@
+	destinationInfo := &oauthv2.DestinationInfo{
+		Config:           destination.Config,
+		DefinitionConfig: destination.DestinationDefinition.Config,
+		WorkspaceID:      destination.WorkspaceID,
+		DefinitionName:   destination.DestinationDefinition.Name,
+		ID:               destination.ID,
+	}
+	yandexUploadManager := &YandexMetricaBulkUploader{
+		destinationInfo: destinationInfo,
+		logger:          logger.NewLogger().Child("batchRouter").Child("AsyncDestinationManager").Child("YandexMetrica").Child("YandexMetricaBulkUploader"),
+	}
+	cache := oauthv2.NewCache()
+	optionalArgs := &oauthv2httpclient.HttpClientOptionalArgs{
+		Logger:    yandexUploadManager.logger,
+		Augmenter: augmenter.YandexReqAugmenter,
+	}
+	originalHttpClient := &http.Client{Transport: &http.Transport{}}
+	// This client is used for uploading data to yandex metrica
+	yandexUploadManager.Client = oauthv2httpclient.NewOAuthHttpClient(
+		originalHttpClient,
+		oauthv2common.RudderFlowDelivery,
+		&cache,
+		backendConfig,
+		augmenter.GetAuthErrorCategoryForYandex,
+		optionalArgs,
+	)
+	return yandexUploadManager, nil
+}
+
+// return a success response for the poll request every time by default
+func (ym *YandexMetricaBulkUploader) Poll(pollInput common.AsyncPoll) common.PollStatusResponse {
+	return common.PollStatusResponse{}
+}
+
+// return a success response for the getUploadStats request every time by default
+func (ym *YandexMetricaBulkUploader) GetUploadStats(UploadStatsInput common.GetUploadStatsInput) common.GetUploadStatsResponse {
+	return common.GetUploadStatsResponse{}
+}
+
+func generateCSVFromJSON(jsonData []byte, goalId string) (string, string, error) {
+	// Define an empty map to store the parsed JSON data
+	var ymMsgs []YandexMetricaMessage
+	inputData := gjson.GetBytes(jsonData, "input").String()
+	err := json.Unmarshal([]byte(inputData), &ymMsgs)
+	if err != nil {
+		return "", "", fmt.Errorf("error while unmarshalling transformed response: %v", err)
+	}
+
+	ymMsgsBody := lo.Map(ymMsgs, func(ym YandexMetricaMessage, _ int) YandexMetricaMessageBody {
+		return ym.Message
+	})
+
+	// Open the CSV file for writing
+	localTmpDirName := fmt.Sprintf(`/%s/`, misc.RudderAsyncDestinationLogs)
+	tmpDirPath, err := misc.CreateTMPDIR()
+	if err != nil {
+		return "", "", fmt.Errorf("error creating tmp dir: %v", err)
+	}
+	folderPath := path.Join(tmpDirPath, localTmpDirName)
+	_, err = os.Stat(folderPath)
+	if os.IsNotExist(err) {
+		folderPath, _ = os.MkdirTemp(folderPath, "")
+	}
+	path := path.Join(folderPath, uuid.NewString())
+	csvFilePath := fmt.Sprintf(`%s.csv`, path)
+	csvFile, err := os.Create(csvFilePath)
+	if err != nil {
+		return "", "", fmt.Errorf("error creating csv file: %v", err)
+	}
+
+	defer csvFile.Close()
+
+	// Create a CSV writer
+	csvWriter := csv.NewWriter(csvFile)
+
+	// Define the header row based on key presence in "message" object
+	firstMsgIdDetails, err := ymMsgsBody[0].Id()
```

Comment:
Could `len(ymMsgsBody)==0` ?

Reply:
No, we are handling that validation in the transformer itself.